_source/_posts/2022-01-19-react-storybook.md

```diff
@@ -0,0 +1,688 @@
+
+```bash
+npm install -E convert-units@2.3.4
+```
+
+Now, in your IDE, create a new file, `src/stories/Converter.jsx`, and fill it with the contents below.
+
+```js
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import * as convert from 'convert-units';
+import { Input, Select } from './Components';
+
+export const Converter = ({measure}) => {
+  const possibilities = convert().possibilities(measure).map((unit) => {
+      const descr = convert().describe(unit);
+      return {
+          value: descr.abbr,
+          description: `${descr.singular} (${descr.abbr})`
+      };
+  });
+
+  const [fromUnit, setFromUnit] = useState(possibilities[0].value);
+  const [toUnit, setToUnit] = useState(possibilities[0].value);
+  const [fromValue, setFromValue] = useState(1);
+  const [toValue, setToValue] = useState(convert(1).from(fromUnit).to(toUnit));
+
+  const updateFromUnit = (event) => {
+    setFromUnit(() => event.target.value);
+    setToValue(() => convert(fromValue).from(event.target.value).to(toUnit));
+  };
+
+  const updateToUnit = (event) => {
+    setToUnit(() => event.target.value);
+    setToValue(() => convert(fromValue).from(fromUnit).to(event.target.value));
+  };
+
+  const updateValue = (event) => {
+    setFromValue(() => event.target.value);
+    setToValue(() => convert(event.target.value).from(fromUnit).to(toUnit));
+  };
+  
+  return <div className="converter">
+      <Select label="From:" options={possibilities} onChange={updateFromUnit}></Select>
+      <Select label="To:" options={possibilities} onChange={updateToUnit}></Select>
+      <Input label="Value:" type="floating-point" onChange={updateValue}></Input>
+      <p>{fromValue} {fromUnit} = {toValue} {toUnit}</p>
+  </div>
+};
+
+Converter.propTypes = {
+  measure: PropTypes.string.isRequired
+};
+
+Input.defaultProps = {
+  measure: 'length'
+};
+```
+
+The component takes a single property called `measure`. The `measure` property specifies the type of units to be converted and can be something like `mass` or `length`. The code for this component then consists of four parts. The first action is to query the `convert-units` library for all the possible unit conversion options. The units are mapped into an array of objects, ready to use with the `Select` component. In the next part, you'll define four state properties. And in the next part, three event handlers are defined. These will react to a change in the user input and update the state accordingly. These event handlers contain the actual calls to the `convert-units` library where the unit conversion happens. Finally, the component is put together from all the parts and returned. You can also create a story for this more complex component with the individual components. Create a file `src/stories/Converter.stories.jsx` and paste the following contents into it.
```

Comment:
```suggestion
The component takes a single property called `measure`. The `measure` property specifies the type of units to be converted and can be something like `mass` or `length`. The code for this component then consists of four parts. The first action is to query the `convert-units` library for all the possible unit conversion options. The units are mapped into an array of objects, ready to use with the `Select` component. In the next part, you'll define four state properties. And in the next part, three event handlers are defined. These will react to a change in the user input and update the state accordingly. These event handlers contain the actual calls to the `convert-units` library where the unit conversion happens. Finally, the component is put together from all the parts and returned. You can also create a story for this more complex component with the individual components. Create a file `src/stories/Converter.stories.jsx` and paste the following contents into it.
```
```suggestion
The component takes a single property called `measure`, which specifies the type of units to be converted and can be something like `mass` or `length`. The code for this component then consists of four parts. The first action is to query the `convert-units` library for all the possible unit conversion options. Units are mapped into an array of objects, ready to use with the `Select` component. In the next part, you'll define four state properties, followed by three event handlers. These will react to a change in the user input and update the state accordingly. These event handlers contain the actual calls to the `convert-units` library where the unit conversion happens. Finally, the component is put together from all the parts and returned. You can also create a story for this more complex component with the individual components. Create a file `src/stories/Converter.stories.jsx` and paste in the following contents.
```
varying sentence length and starting words, instead of repeating "The" - reduces monotony, increases readability